### ISSUE TITLE
[FEATURE] Make Transport Request call async

### ISSUE THREAD

**Issue by owaiskazi19** [opensearch-project/opensearch-sdk#17]
### Is your feature request related to a problem?
Current Transport Request calls to OpenSearch are synchronous, a better way will be to make them asynchronous if in them future we want to make parallel request.

### What solution would you like?
Use ActionListener pattern present in OpenSearch to make async calls.

### What alternatives have you considered?
_A clear and concise description of any alternative solutions or features you've considered._

### Do you have any additional context?
_Add any other context or screenshots about the feature request here._

**Comment by minalsha** [comment]
Can you share the status on this @owaiskazi19  ?

**Comment by dbwiddis** [comment]
Most of the original request has been handled via other issues.  Updated the initial comment to describe the one remaning task to do, (a)synchronizing the initialization sequence.

**Comment by dbwiddis** [comment]
> when the error occurs during register request execution, should the ExtensionsRunner manually `throw` an exception so `AcknowledgedResponseHandler` will handle it, or is it primarily based on the status that lies inside `AcknowledgedResponse`? 

You'll need to pass a `CompletableFuture` to the `AcknowledgedResponseHandler` (you can create a new constructor with that Future as a parameter).  Then in the `handleResponse()` block you'll complete the future; in the `handleException()` block you'll complete the future exceptionally.  So some very rough pseudocode:
 - in each `sendFooRequest()` method (most in `ExtensionsRunner` but not all), Instantiate a CompletableFuture 
 - Instantiate an AcknowledgedResponseHandler for each request using that future as a parameter
 - change the return of those methods from `void` to returning the future
 - in the `ExtensionsInitRequestHandler` collect all of those futures and wait for all of them to complete

> Are there any examples in the repository where I can see similar functionality to make it more straightforward for me?

I'm not sure we have done this other than anonymous subclasses [like here](https://github.com/opensearch-project/OpenSearch/blob/73a2279fdc3b432f9ad9f25a773c0c987ccb56f4/server/src/main/java/org/opensearch/extensions/rest/RestSendToExtensionAction.java#L128-L162) but the concept should be similar (just imagine the future as a parameter internal to the handler) that gets returned.

**Comment by dbwiddis** [comment]
Hey @qchimera are you actively working on this?  No worries if you are and need a bit more time, just let me know.

**Comment by dbwiddis** [comment]
Great question.

When we wrote that method we essentially made it synchronous, so it served as a natural stopping point for the initialization.  It's your choice whether to rewrite it to be asynchronous the same as everything else (probably my preference) or to leave it as is.

Note that it must complete before the code using its result.

**Comment by dbwiddis** [comment]
> I re-read it and it sounds horrible, hope you'll understand me xD

Based on that disclaimer I skipped reading to independently see what needed to be done, let's see if it matches yours:
 - you'll be removing the `awaitResponse()` code and directly returning the `CompletableFuture<EnvironmentSettingsResponse>` to the init request handler.
 - You'll add that one to the `allOf()` call which must complete before extracting the `settings` from that future to use in the follow-on code.

Now I read your description and that looks like what you said! :)

**Comment by dbwiddis** [comment]
I'm still debugging something on #512 so don't expect it until this weekend.

I suggest you do a few things:
 - stop using your `main` branch for features.  😁  You're not the only one to make mistakes like this.   Always get a fresh copy of `main` and create a new branch to start working, and rebase with main often.
 - to fix this problem use interactive `git rebase`.  DO NOT DO THIS ON MAIN.  Create a branch (`git checkout -b newbranchnamehere`).
   - `git rebase -i HEAD~9` will show you the last 9 commits (not sure if that's too few or too many).  Delete the commits not associated with the current PR. Then you can push this and create a new PR.

GIven that your open PR from `main` is blocking you from properly using git, I'll prioritize getting that merged soon, probably this weekend.
